### PR TITLE
Adding greedy pixel pick and fixing nodata error for percentiles

### PR DESCRIPTION
**Pull request by richpsharp**
Adding the `greedy_pixel_pick` routine to pygeoprocessing and fixing an issue for percentile calculation that would cause a crash if rasters had undefined nodata or nan values in the pixels.

### CODE REVIEW

File: tests/test_geoprocessing.py [hunk 1]
```diff
@@ -4317,3 +4418,30 @@ def test_convolve_with_byte_kernel(self):
...
+
+        area_path = os.path.join(test_dir, 'area_raster.tif')
+        _array_to_raster(
+            numpy.ones((n_pixels, n_pixels), numpy.float32), target_nodata, area_path)
+
+        pygeoprocessing.greedy_pixel_pick_by_area(
+            (value_path, 1), (area_path, 1),
+            [1, 10, 100], test_dir, output_prefix='test_')
+
+        self.assertTrue(os.path.exists(os.path.join(
```
Comment (beckyck) [inline tests/test_geoprocessing.py]: could you test to see if the rasters and the values in the table are correct?

**Review by beckyck — CHANGES_REQUESTED** [review]
Just a few minor issues!

File: src/pygeoprocessing/geoprocessing.py [hunk 2]
```diff
@@ -4007,3 +4007,110 @@ def _create_latitude_m2_area_column(lat_min, lat_max, n_pixels):
         _m2_area_of_wg84_pixel(pixel_size, lat)
         for lat in reversed(center_lat_array)]).reshape((n_pixels, 1))
     return area_array
+
+
+def greedy_pixel_pick_by_area(
+        base_value_raster_path_band, area_per_pixel_raster_path_band,
+        selected_area_report_list, output_dir, output_prefix=None,
+        ffi_buffer_size=2**11, heap_buffer_size=2**28):
```
[the remark this answers is not in the document]
Comment (richpsharp) [reply src/pygeoprocessing/geoprocessing.py]: fixed in 98da3cafaff5c44f0cf94d2bec78df84bfa536c1

File: src/pygeoprocessing/geoprocessing.py [hunk 3]
```diff
@@ -4007,3 +4007,110 @@ def _create_latitude_m2_area_column(lat_min, lat_max, n_pixels):
...
+
+    sorted_area_sum = numpy.cumsum(sorted_area)
+    value_sum = numpy.cumsum(base_array[valid_mask][sort_indexes])
+
+    area_threshold_index_array = numpy.searchsorted(
+        sorted_area_sum, selected_area_report_list, side='right')
+
+    os.makedirs(output_dir, exist_ok=True)
+    table_path = os.path.join(
+        output_dir, f'{output_prefix}greedy_pixel_pick_result.csv')
```
[the remark this answers is not in the document]
Comment (richpsharp) [reply src/pygeoprocessing/geoprocessing.py]: fixed in 98da3cafaff5c44f0cf94d2bec78df84bfa536c1

File: tests/test_geoprocessing.py (hunk 1 above)
Comment (richpsharp) [reply tests/test_geoprocessing.py]: fixed in 98da3cafaff5c44f0cf94d2bec78df84bfa536c1

**Comment by richpsharp** [comment]
Hi @beckyck thanks for the review I think i addressed all those issues you saw!

**Review by beckyck — APPROVED** [review]

**Comment by richpsharp** [comment]
thanks @beckyck !

**Pull request merged**